### PR TITLE
Boost: Don't build unit-test by default & support CMAKE_OSX_ARCHITECTURES

### PR DESCRIPTION
**Pull request by hjmallon**
* I've checked this [Git style guide](https://0.readthedocs.io/en/latest/git.html). **[Yes]**
* I've checked this [CMake style guide](https://0.readthedocs.io/en/latest/cmake.html). **[Yes]**
* My change will work with CMake 3.2 (minimum requirement for Hunter). **[Yes]**
* I will try to keep this pull request as small as possible and will try not to mix unrelated features. **[Yes]**

---

My intent was to fix making universal Boost libraries (arm64/x86_64), but I also noticed that unit-tests was building by default (x86_64) and needed to be removed.

@Bjoe may be interested?

### CODE REVIEW

**Comment by hjmallon** [comment]
I have updated the PR:
* I needed to update the `PACKAGE_INTERNAL_DEPS_ID`
* I found that Boost 1.75.0 has a new `json` built lib which needed adding - hope I have done this right

**Comment by Bjoe** [comment]
Looks good for me but all the ios builds are failing with `[hunter ** FATAL ERROR **] ABI not detected for C compiler` ? Should this not fixed with your PR?

Also....
* Should we add an example for Boost json?
* Should we verify if unit-test also builds for the other platforms?

Not sure. @rbsheth ?

**Comment by hjmallon** [comment]
> Looks good for me but all the ios builds are failing with [hunter ** FATAL ERROR **] ABI not detected for C compiler ? Should this not fixed with your PR?

@Bjoe I don't know much about iOS. Should I just disable the `CMAKE_OSX_ARCHITECTURES` thing for iOS. Is that just `if (NOT IOS)` or something?

>
> Also....
>
>    Should we add an example for Boost json?

Not all the Boost sub-libs have examples at the moment, but I could add one easy enough (I think, never used the Boost json - I added it so that it was disabled by default).

>    Should we verify if unit-test also builds for the other platforms?

Not sure what you mean, do you remember why you added the `UNTIL 1.74.0`? Was Boost unit-test meant to be deprecated or something?

**Comment by Bjoe** [comment]
Btw. you should add json in  `cmake/projects/Boost/generate.sh` also and I'm already thinking about, why I added `until 1.74.0` on `hunter_append_component(${common_args} COMPONENT test            SINCE 1.21.0 UNTIL 1.74.0)`  ... hm.
I'm also confused about your PR: "Don't build unit-est by default". What you mean with that? With this "UNTIL 1.74.0" it will not build test at all for 1.74.0.

**Comment by Bjoe** [comment]
> > Looks good for me but all the ios builds are failing with [hunter ** FATAL ERROR **] ABI not detected for C compiler ? Should this not fixed with your PR?
> 
> @Bjoe I don't know much about iOS. Should I just disable the `CMAKE_OSX_ARCHITECTURES` thing for iOS. Is that just `if (NOT IOS)` or something?
> 
Ah ok, you fixed the "MacOS" build. Ok this info I missed.
>
> > Also....
> > Should we add an example for Boost json?
> 
> Not all the Boost sub-libs have examples at the moment, but I could add one easy enough (I think, never used the Boost json - I added it so that it was disabled by default).
>
Again not sure. @rbsheth ? 
>
> > Should we verify if unit-test also builds for the other platforms?
> 
> Not sure what you mean, do you remember why you added the `UNTIL 1.74.0`? Was Boost unit-test meant to be deprecated or something?

Also not sure yet ... I'm try to find out why I added this.

**Comment by hjmallon** [comment]
> Btw. you should add json in cmake/projects/Boost/generate.sh

Already done 😄 

> I'm also confused about your PR: "Don't build unit-est by default". What you mean with that? With this "UNTIL 1.74.0" it will not build test at all for 1.74.0.

So I am testing with Boost 1.75.0. One of the uses of that list of libraries is for Hunter to do `./b2 ... --without-json ...`, so it has to know all the libraries in order to disable all the libraries. Without it sometimes you get:

```
Component configuration:

    - atomic                   : not building
    - chrono                   : not building
    - container                : not building
    - context                  : not building
    - contract                 : not building
    - coroutine                : not building
    - date_time                : not building
    - exception                : not building
    - fiber                    : not building
    - filesystem               : not building
    - graph                    : not building
    - graph_parallel           : not building
    - headers                  : building
    - iostreams                : not building
    - json                     : building
    - locale                   : not building
    - log                      : not building
    - math                     : not building
    - mpi                      : not building
    - nowide                   : not building
    - program_options          : not building
    - python                   : not building
    - random                   : not building
    - regex                    : not building
    - serialization            : not building
    - stacktrace               : not building
    - system                   : not building
    - test                     : not building
    - thread                   : not building
    - timer                    : not building
    - type_erasure             : not building
    - wave                     : not building
```

**Comment by hjmallon** [comment]
I have added a Boost::json example

**Comment by Bjoe** [comment]
> > Btw. you should add json in cmake/projects/Boost/generate.sh
> 
> Already done smile
>
:+1: 
> 
> > I'm also confused about your PR: "Don't build unit-est by default". What you mean with that? With this "UNTIL 1.74.0" it will not build test at all for 1.74.0.
> 
> So I am testing with Boost 1.75.0. In this case unit-test is build by default during the Boost headers build and the libraries end up in the Hunter Install tree.
> 
Great, Good finding !

File: cmake/projects/Boost/ci/matrix.json [hunk 1]
```diff
@@ -126,6 +126,23 @@
...
+{ "example": "Boost-json", "toolchain": "ninja-vs-15-2017-win64-cxx17",                    "os": "windows-2016", "python": "3.8", "script": "build.cmd" },
+{ "example": "Boost-json", "toolchain": "nmake-vs-15-2017-win64-cxx17",                    "os": "windows-2016", "python": "3.8", "script": "build.cmd" },
+{ "example": "Boost-json", "toolchain": "vs-16-2019-win64-cxx17",                          "os": "windows-2019", "python": "3.8", "script": "build.cmd" },
+{ "example": "Boost-json", "toolchain": "vs-15-2017-win64-cxx17",                          "os": "windows-2016", "python": "3.8", "script": "build.cmd" },
+{ "example": "Boost-json", "toolchain": "vs-14-2015-sdk-8-1",                              "os": "windows-2016", "python": "3.8", "script": "build.cmd" },
+{ "example": "Boost-json", "toolchain": "mingw-cxx17",                                     "os": "windows-2016", "python": "3.8", "script": "build.cmd" },
+{ "example": "Boost-json", "toolchain": "msys-cxx17",                                      "os": "windows-2016", "python": "3.8", "script": "build.cmd" },
+
```
Comment (Bjoe) [inline cmake/projects/Boost/ci/matrix.json:145]: Unfortunately, this is to much for the github builds. I think, only 256 jobs are allowed for one "action". Maybe you add some of them and remove some other ....

Comment (Bjoe) [reply cmake/projects/Boost/ci/matrix.json:145]: See also `CI / set_status (pull_request) Skipped `

Comment (hjmallon) [reply cmake/projects/Boost/ci/matrix.json:145]: I have commented out a few of the lines and it seems to run now. Maybe we should thin out the Boost builds a bit in general in a separate change?

Comment (rbsheth) [reply cmake/projects/Boost/ci/matrix.json:145]: That sounds like a plan

**Review by rbsheth — APPROVED** [review]
Thanks!!

**Comment by rbsheth** [comment]
Released: https://github.com/cpp-pm/hunter/releases/tag/v0.23.298

**Pull request merged**